File: app/src/main/java/com/ebsoft/watchlist/ui/base/BaseViewModel.java
package com.ebsoft.watchlist.ui.base;

import android.arch.lifecycle.ViewModel;

import com.ebsoft.watchlist.data.DataManager;

import java.lang.ref.WeakReference;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * Created by endre on 07/09/18.
 */

public abstract class BaseViewModel<N> extends ViewModel {

    private WeakReference<N> mNavigator;
    protected final DataManager mDataManager;
    protected final CompositeDisposable mCompositeDisposable;

    public BaseViewModel(DataManager DataManager) {
        this.mDataManager = DataManager;
        this.mCompositeDisposable = new CompositeDisposable();
    }

    public DataManager getDataManager() {
        return mDataManager;
    }

    public void addDisposable(Disposable disposable) {
        mCompositeDisposable.add(disposable);
    }

    public N getNavigator() {
        return mNavigator.get();
    }

    public void setNavigator(N navigator) {
        this.mNavigator = new WeakReference<>(navigator);
    }

    @Override
    protected void onCleared() {
        mCompositeDisposable.dispose();
        super.onCleared();
    }
}

File: app/src/main/java/com/ebsoft/watchlist/ui/base/BaseViewModel.kt
package com.ebsoft.watchlist.ui.base

import android.arch.lifecycle.ViewModel
import com.ebsoft.watchlist.data.DataManager
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import java.lang.ref.WeakReference
import kotlin.coroutines.CoroutineContext

/**
 * Created by endre on 07/09/18.
 */

abstract class BaseViewModel<N>(val dataManager: DataManager) : ViewModel(), CoroutineScope {

    private var mNavigator: WeakReference<N>? = null

    private val mCompositeDisposable: CompositeDisposable = CompositeDisposable()

    private val job = Job()

    override val coroutineContext: CoroutineContext get() = job + Dispatchers.Main

    var navigator: N
        get() = mNavigator?.get()!!
        set(navigator) {
            this.mNavigator = WeakReference(navigator)
        }

    fun addDisposable(disposable: Disposable) {
        mCompositeDisposable.add(disposable)
    }

    override fun onCleared() {
        mCompositeDisposable.dispose()
        job.cancel()
        super.onCleared()
    }
}
